File: application/backend-credit-card-enrollment/backend-c#/Common/Query/QueryController.cs
using CreditCardEnrollment.Common.Projection;

namespace CreditCardEnrollment.Common.Query;

public class QueryController {
    private readonly MongoTransactionalProjectionOperator _mongoTransactionalProjectionOperator;
    private readonly ILogger<QueryController> _logger;

    public QueryController(
        MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator,
        ILogger<QueryController> logger
    ) {
        _mongoTransactionalProjectionOperator = mongoTransactionalProjectionOperator;
        _logger = logger;
    }

    protected object ProcessQuery(Query query, QueryHandler queryHandler) {
        try {
            _logger.LogDebug("Starting to process query: {QueryType}", query.GetType().Name);
            _mongoTransactionalProjectionOperator.StartTransaction();
            var result = queryHandler.HandleQuery(query);
            _mongoTransactionalProjectionOperator.CommitTransaction();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();

            _logger.LogDebug("Successfully processed query: {QueryType}", query.GetType().Name);
            return result;
        } catch (Exception ex) {
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            _logger.LogError("Exception in ProcessQuery: {0}, {1}", ex.Message, ex.StackTrace);
            throw new Exception($"Failed to process query: {ex.Message}", ex);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/common/query/QueryController.ts
import { MongoTransactionalProjectionOperator } from '../projection/MongoTransactionalProjectionOperator';
import { log } from '../util/Logger';
import { QueryHandler } from './QueryHandler';
import { Query } from './Query';

export class QueryController {
    constructor(
        private readonly mongoTransactionalProjectionOperator: MongoTransactionalProjectionOperator
    ) {}

    protected async processQuery(query: Query, queryHandler: QueryHandler): Promise<unknown> {
        try {
            log.debug(`Starting to process query: ${query.constructor.name}`);
            await this.mongoTransactionalProjectionOperator.startTransaction();
            const result = await queryHandler.handleQuery(query);
            await this.mongoTransactionalProjectionOperator.commitTransaction();
            await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();

            log.debug(`Successfully processed query: ${query.constructor.name}`);
            return result;
        } catch (error) {
            await this.mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
            log.error(`Exception in ProcessQuery: ${error}`, error as Error);
            throw new Error(`Failed to process query: ${error}`);
        }
    }
}